### ISSUE TITLE
[bndtools debug] Add proper OSGi-aware runtime source lookup

### ISSUE THREAD

**Issue by kriegfrj** [bndtools/bnd#4029]
In the wake of #4011, the source lookup in Bndtools is a lot better but it is still not 100% because it is built on the basic built-in Eclipse JDT Java source lookup. This source lookup assumes a flat classpath and is not classloader aware, eg:

1. The `ISourceContainers` are asked to search for the source for a class by name only, with no hint as to which bundle it might belong to.
2. The `CompositeSourceContainer` (which is used in #4011) simply searches its children sequentially for the first match - or, if you've allowed duplicate sources, it will return all the matches and then allow the user to select one.

So if you have multiple versions of a class in `-runbundles` in different bundles, you are likely to run into issues with source selection.

The solution to this is for `BndDependencySourceLookupDirector` to override the default behaviour. As far as I can understand, the debug source lookup is in three parts in Eclipse:

1. A class that maps from a class from Eclipse's debug model (eg, IJavaStackFrame) to a String that the source lookup containers (`ISourceContainer`) will understand. This is the `ISourceLookupParticipant`
2. The container classes that take this String and find the corresponding source file.
3. The `ISourceLookupDirector` that coordinates all of the above.

For completeness, there is also `ISourcePathLocator` (which is legacy and a superinterface of `ISourceLookupDirector`), `ISourcePathComputer` (which I'm not 100% sure what it does but I think it is used by the `DefaultSourceContainer` to generate the `ISourceContainer` instances). And there is `IRuntimeClasspathEntry`, which seems to be used as an intermediary to compute the source containers. I don't think we need to worry about `ISourcePathComputer` or `ISourcePathComputer` as yet (though a deeper understanding might change that opinion) - we can do everything we need in `BndDependencySourceLookupDirector`. To see how this could be implemented, we can use `PDESourceLookupDirector` (https://github.com/eclipse/eclipse.pde.ui/blob/master/ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/sourcelookup/PDESourceLookupDirector.java) as a reference. 

Basically, the idea is to override `getSourceElement()` and `findSourceElement()` so that it does a bundle-aware search - actually looking in to the IJavaStackFrame to find the classloader that loaded the class and act accordingly. This bundle information can be encoded in the search result and passed in to the containers so that they look in the right bundle. This can also be used to optimisee the search (eg, if it was the system classloader or a non-bundle classloader, then we don't need to look in `-runbundles` at all). Things like fragments and `Bundle-Classpath` complicate this a bit, but nothing that couldn't be handled.

**Comment by pkriens** [comment]
Men, you're doing a superhuman effort here. You seem to be deep in the well that I always dislike so much, trying to figure out how the Eclipse code is supposed to be used :-(

To understand what we're talking about, I put it in an Alloy model, the picture might also help you a bit.

[image](https://user-images.githubusercontent.com/200494/81267071-228d0880-9046-11ea-8373-7e7670aafde8.png)

**Comment by kriegfrj** [comment]
Thanks Peter - where was this diagram when I needed it last week??? :wink:

In the Bndtools implementation we seem to have bypassed the `ISourcePathComputer` - the `BndDependencySourceLookupDirector` directly builds the root `ISourceContainer` (`BndDependencySourceContainer`), which in turn builds its children.

**Comment by pkriens** [comment]
I think many constructs in Bndtools were written in the dark. I really like what you're doing but it seems a lot of work. This exercise made it a lot clearer to me why Eclipse is losing ground due to its myriad of really bad plugins. If you look for implementations of these classes, I find surprisingly few open source projects that implement them.

**Comment by kriegfrj** [comment]
Yes, and the diagram doesn't even tell the whole story... what about IRuntimeClasspathEntry and IClasspathEntry? It has been some work, but when i add up the time I've lost because this stuff isn't working it's probably not so bad (not having the Eclipse sources handy when trying to write the integration tests was painful). And to be fair  i haven't tried the equivalent exercise under IntelliJ.

Btw, the comment about us bypassing ISourcePathComputer was not necessarily a criticism of Bndtool's design - i don't think it needs it. It seems that is more suited to a more traditional Java project model.

**Comment by kriegfrj** [comment]
I started having a look at this. Unfortunately, some of the key code that I will need (eg, IEvaluationEngine) is in an embedded JAR in a bundle (`org.eclipse.jdt.debug`) that makes use of `Bundle-ClassPath`, so the compiler in Bndtools cannot see it at the moment. In other words, this has brought me back to #2415 and bndtools/bndtools#1871. Hopefully now with two more years of Bnd/tools experience under my belt I can have a look at that now.

**Comment by pkriens** [comment]
Could we make the source lookup lazy? We do not automatically download the sources. The way it is currently done means that we will always download the sources, and actually right in the middle of the build when we initialize the classpath. 

Could make a delegated IClassPathEntry that would delegate to the real guy but fetches the source code when the getSourcePath() is called. We could actually download the source via a Promise and use the promise in the unlikely case it is used.

I have a customer who has no sources for a lot of maven projects. Currently, it always looks up the sources and goes online, only to get a 404. I also talked to BJ to see if we could build a 404 cache in the client but he wasn't too keen on this.

Tell me if you want me to make a proposal for this.

**Comment by kriegfrj** [comment]
> Could we make the source lookup lazy? We do not automatically download the sources. The way it is currently done means that we will always download the sources, and actually right in the middle of the build when we initialize the classpath.
>
> Could make a delegated IClassPathEntry that would delegate to the real guy but fetches the source code when the getSourcePath() is called. We could actually download the source via a Promise and use the promise in the unlikely case it is used.
> 
> I have a customer who has no sources for a lot of maven projects. Currently, it always looks up the sources and goes online, only to get a 404. I also talked to BJ to see if we could build a 404 cache in the client but he wasn't too keen on this.
> 
> Tell me if you want me to make a proposal for this.

This is a good idea. I think it is feasible.

This issue was originally intended to deal with source lookup at launch time rather than build time (ie, `ISourceContainer`/`ISourceLookupDirector` rather than `IClasspathEntry`). My implementation initiates download for the source bundles during the launch initialisation but doesn't wait for it to finish - instead it uses a Promise (via DownloadListener). That way it doesn't slow down the launch but the sources should be ready when you want them.

However, the same concept could probably be applied to `IClassPathEntry` so that the download of sources doesn't happen synchronously on the build thread. That way the sources could be ready when you need them (making code navigation more responsive in the editor) but it shouldn't not slow down the build too much.

I'm happy for you to take this and run with it, as I've got plenty of other things I should be doing... :smile:
